### DISCUSSION TITLE
Possible To Integrate Raspberry Pi UART TX as WS2812B Driver?

### DISCUSSION THREAD

**Discussion by jliriano2000** [hyperion-project/hyperion.ng#1789] [Ideas]
I followed this [guide](https://adanbucio.net/2017/07/31/how-to-drive-neopixels-from-a-pc-without-a-microcontroller/) and was able to solve two problems with one stone: 1) Level shift the data signal to 5v and 2) Not rely on the PWM or SPI peripheral. 

I've confirmed that you can drive an arbitrary length of WS2812 with the inverted UART TX pin using this method. Is there already a way to use Hyperion with this set-up? If not, I'd be interested in working on integrating it in as a contributor.

**Comment by Lord-Grey** [comment]
We have lately added support for ftdi devices which I think is the scenario you are describing.
See https://github.com/hyperion-project/hyperion.ng/pull/1746

You can install a nightly build and try it out.

https://releases.hyperion-project.org/

Go to Linux -> easy install and tick nightly to get the install command.

Note: It is not available for Windows yet, but a PR is in the pipeline…

**Reply by jliriano2000** [reply]
@Lord-Grey 
I took a closer look at the PR. It seems like you are using the GPIO bits of an FTDI device to make a SPI device instead of using the UART TX pin. 

The FTDI UART TX pin can drive the WS2812 if one: 
1) Sets the FTDI part to  2.4 or 3 MBaud, 7 bits, 1 Stop, No parity
2) Inverts the TX pin by updating the EEPROM or using a transistor and a pull up like in the [guide](https://adanbucio.net/2017/07/31/how-to-drive-neopixels-from-a-pc-without-a-microcontroller/). 

This can be abstracted to any UART device (not just FTDI) that supports the configuration items above. 
There are already LED devices that use UART as a programming interface under /leddevice/led_serial 

It seems it would be straightforward to add a .cpp and .h file for the WS2812 device in that directory.

**Reply by Lord-Grey** [reply]
Hi
I did not implement the Ftdi device and I have to admit, I only remembered USB to Serial device...
Thanks for outlining that the output is a SPI signal.

For the pure serial scenario, have you looked at the HIDRaw device?
I have no background on those, but they look like USB to serial ones...

**Reply by jliriano2000** [reply]
@Lord-Grey That does not seem to be a solution either. I think I will take it on myself as a side project to add the serial functionality if it does not exist. If successful, I can open a PR. How may one join the team of contributors?

**Reply by Lord-Grey** [reply]
For testing, you could just grab one of the [serial devices](https://github.com/hyperion-project/hyperion.ng/tree/master/libsrc/leddevice/dev_serial) and update the code along the UART way of working. 
if it works, I can support you getting it to a consistent new device.
That way you are not required to have everything defined from UI to backend code to explore feasibility.

The tricky design part might be, if we want to mix ftdi control code with the serial output code (if we want using the ftdi invert capability).

Edit: Looks like the ftdi device can be generally configured with invention using the FT_PROG program. I will check this out, if I find time. That would avoid any variation in code at the beginning…

To get a contributor just issue a PR.

**Comment by jliriano2000** [comment]
It looks like this was merged into master. If I compile master, I should be
good to go!
Thanks for your help!

**Comment by jliriano2000** [comment]
@Lord-Grey I have an update: 

So I was incorrect about driving WS2812 LED's with the inverted pin of a FT232R chip. The timing is close but not close enough to drive the LEDs correctly. Also, the -R versions do not support turning the GPIO into SPI busses. So sad 👎 

The Raspberry PI UART TX pin, however, IS capable of driving WS2812 LED's when configured for 3 MBaud, 7 bits, 1 Stop bit.  The only draw-back is that the pin has to be inverted and level-shifted with hardware. 

I downloaded the source and hacked the LedDeviceTpm2.cpp and .h files and the ProviderRs232.cpp and .h files to hard-code the serial configuration. I then recompiled and ran the server. The result is in the attached video. 

You will notice the display glitch for a frame and display random noise.  This may be caused by the serial stream getting interrupted by something else running on the Pi, triggering the WS2812 latch time.  Is there a remedy for this ? 


Video:
https://github.com/user-attachments/assets/b944ba0e-93ee-433a-b1f7-ee9b03ffbcbd

**Comment by Lord-Grey** [comment]
@jliriano2000 Thanks for taking the time doing an implementation and testing it out.
Have you tried reducing the fps rate and/or increasing the latch time? Not that the serial device is flooded with too many updates.

In addition, is there a chance that you share the code, that I have a try with an ftdi board I have in case I find some time giving it a try too?
You could PM it to me via the Hyperion forum or via Discord.

**Reply by jliriano2000** [reply]
@Lord-Grey  I will play with the settings tonight to see if I can eliminate the glitch, but the more I think about it, the more I am sure that the byte stream is getting interrupted by some higher priority process in the Raspberry Pi.  There cannot be a delay between bytes, or the LED's will latch prematurely. An FTDI SPI device might be more gracious since it has its own internal buffers which are immune to whatever is going on in the Pi. 

I have attached the python scripts I have used to investigate using serial ports to drive the WS2812.  The serial port is the -c argument (ie, -c /dev/serial0 on a raspberry pi, -c /dev/tty.usbserial-xxx on a Mac)

For using an FTDI board, make sure to invert the logic on the TX pin and set the I/O voltage to 5V. I used the [FTPROG](https://ftdichip.com/utilities/#ft_prog) Windows program for this. 

Part of the problem with the FTDI UART board is the resolution of the baud rate setting. Somewhere between 2 MBaud and 3 MBaud is a magic speed that aligns with the WS2812 timing.  The board does not support enough speeds between those two values. At least the -R type I used did not. 

On a raspberry pi, however, the 3 MBaud setting on its internal UART works like a charm. 


[pythonSource.zip](https://github.com/user-attachments/files/17401252/pythonSource.zip)

**Reply by Lord-Grey** [reply]
Thank you.
I was looking for the cpp code of the hacked Hyperion files.

**Reply by jliriano2000** [reply]
Ah sorry! I will send those when I get home tonight.

**Reply by jliriano2000** [reply]
@Lord-Grey Here are the modifications I made to the serial files:
[Serial Mods.zip](https://github.com/user-attachments/files/17404046/Serial.Mods.zip)

I've messed with the settings as promised and still can't get the glitching to go away.  Please take a look at the code to see if there is something there causing it. 

Some patterns are more glitchy than others. The simple led test that does a walking one down the line of LED's does not glitch at all.
